### ISSUE TITLE
remove the /dev/urandom backend from twisted.python.randbytes

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4530]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)| @zooko reported|
|-|-|
|Trac ID|trac#4530|
|Type|enhancement|
|Created|2010-06-30 04:05:29Z|
|Branch|https://github.com/twisted/twisted/tree/remove-devurandom-4530|

```twisted.python.randbytes``` uses ```os.urandom()``` and if that fails then reads from ```/dev/urandom```. However, ```os.urandom()``` itself always reads from ```/dev/urandom```, so randbytes can leave out the part about reading from ```/dev/urandom``` itself.

See the ```def urandom``` at the end of ```Lib/os.py```:

http://svn.python.org/view/python/branches/release24-maint/Lib/os.py?view=markup

Note that ```os.urandom``` was introduced in Python 2.4. There would have been a use for Twisted reading from ```/dev/urandom``` itself back when Twisted supported Python 2.3.

By the way I just audited the os.urandom implementation (in current trunk ~= Python 2.7) and found a vulnerability in the VMS implementation -- http://bugs.python.org/issue9123 -- but otherwise didn't see any problem with it. Twisted's implementation is better (don't catch arbitrary exception type, don't attempt to handle short reads). Maybe we should submit a patch to Python to improve their implementation at the same time as we remove ours?

Attachments:

* [fix.urandom.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/344/344f48cf82c0521b348840d80fb7d20d07157626/b59108bafe0211e431d84953f3a176540c15da58.patch) (380 bytes) - added by zomux on 2012-03-31 15:57:59Z - 
* [remove._fileUrandom.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/344/344f48cf82c0521b348840d80fb7d20d07157626/f00f2533a26476eb1c303582206a6a7b4a5cdcbf.patch) (3244 bytes) - added by zomux on 2012-03-31 23:59:06Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__4530 4530
type__enhancement enhancement
reporter__zooko zooko
priority__normal normal
milestone__ 
branch__branches_remove_devurandom_4530 branches/remove-devurandom-4530
branch_author__itamarst itamarst
status__closed closed
resolution__fixed fixed
component__core core
keywords__security security
time__1277870729000000 1277870729000000
changetime__1333310887000000 1333310887000000
version__None None
owner__itamar itamar
cc__zooko@... cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="zomux's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|zomux commented|
|-|-|

just removed '/dev/urandom' from random sources at line 41

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)|@dreid set owner to @dreid<br>@dreid set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_6"></a>|@dreid removed owner<br>@dreid set status to `new`|
|-|-|

Thanks for working on this zomux.  It's always great to see new contributors.

Address these and submit again for review.

 * Empty tuple syntax is ```()``` not ```(,)```, the latter is invalid syntax so twisted.python.randbytes is no longer importable.
 * ```twisted.test.test_randbytes.ConditionalSecureRandomTestCase.test_fileUrandom``` now fails.

I'll attempt to solicit some feedback on this issue from other developers.  Your input is of course also appreciated so feel free to respond here, I just wouldn't bother doing any work yet.

 * It may be reasonable to remove the ```_fileUrandom``` implementation and the code that tries ```_osUrandom``` & ```_fileUrandom```
  * However, this might require deprecation first, since though ```_fileUrandom``` is private ```randomSources``` is public.
   * of course the whole class could be refactored to a set of module-level functions once you remove the ```randomSources``` state.

Great work.  Thanks again for contributing to twisted.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_7"></a>|@itamarst commented|
|-|-|

Given that reading from given list of files only ever happened after trying `os.urandom`, and modern Python versions have had that for more than 5 years, I don't think anyone's actually ever used _fileUrandom code path so.

So:

  1. Fix the first two issues dreid mentioned above.
  2. Remove the _fileUrandom method and its unit test. 
  3. Add comment above randomSources indicating it is no longer used by anything and will eventually be removed.
  4. Add a news file, as explained here (http://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles) -- "svn add"ing the file will make sure it shows up in svn diff. 
  5. Make sure the tests run before submitting a patch.

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="zomux's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|zomux commented|
|-|-|

1.removed _fileUrandom method from twisted.python.randbytes.RandomFactory 
   and only use _osUrandom method to call os.urandom function .

2.removed unit test of it

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_9"></a>|@itamarst commented|
|-|-|

(In [34051]) Branching to 'remove-devurandom-4530'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_10"></a>|@itamarst set owner to @itamarst|
|-|-|

Buildbot is happy, other than the usual Windows issues:
http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/remove-devurandom-4530

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_11"></a>|@itamarst set status to `closed`|
|-|-|

(In [34054]) Merge remove-devurandom-4530: Get rid of dead code for reading from /dev/urandom.

Author: zomux
Review: itamar
Fixes: [#4530](https://github.com/twisted/twisted/issues/4530)

os.urandom has been available since Python 2.4, and since the code using it was first, the code path that tried to read from the /dev/urandom file would never be reached in practice.
